src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -104,32 +105,38 @@ def _make_candidate_from_link(
             return ExtrasCandidate(base, extras)
         return base
 
-    def make_candidate_from_ican(
-        self,
-        ican,  # type: InstallationCandidate
-        extras,  # type: Set[str]
-        parent,  # type: InstallRequirement
-    ):
-        # type: (...) -> Candidate
-        dist = self._installed_dists.get(ican.name)
-        should_use_installed_dist = (
-            not self._force_reinstall and
-            dist is not None and
-            dist.parsed_version == ican.version
+    def iter_found_candidates(self, ireq, extras):
+        # type: (InstallRequirement, Set[str]) -> Iterator[Candidate]
+        name = canonicalize_name(ireq.req.name)
+        if not self._force_reinstall:
+            dist = self._installed_dists.get(name)
+        else:
+            dist = None
+
+        found = self.finder.find_best_candidate(
+            project_name=ireq.req.name,
+            specifier=ireq.req.specifier,
+            hashes=ireq.hashes(trust_internet=False),
         )
-        if not should_use_installed_dist:
-            return self._make_candidate_from_link(
+        for ican in found.iter_applicable():
+            if dist is not None and dist.parsed_version == ican.version:
```

Comment:
nit: Maybe rename `dist` as `installed_dist` - the point of this check wasn't obvious to me when the name is just `dist`. I think just renaming the variable would make that check clearer. Or you could add a comment :-)

Reply:
Updated